File: TP1_Crypto/Analyse coincidance/indice_coincidence.py
from PyQt5.QtWidgets import QApplication, QWidget, QVBoxLayout, QLabel, QTextEdit, QPushButton
import sys

def calculer_indice_coincidence(texte):
    frequences = {chr(i): 0 for i in range(ord('a'), ord('z') + 1)}
    total_lettres = 0
    
    for c in texte.lower():
        if 'a' <= c <= 'z':
            frequences[c] += 1
            total_lettres += 1
    
    if total_lettres <= 1:
        return 0.0
    
    ic = sum(f * (f - 1) for f in frequences.values()) / (total_lettres * (total_lettres - 1))
    return ic

class ICApp(QWidget):
    def __init__(self):
        super().__init__()
        self.initUI()

    def initUI(self):
        layout = QVBoxLayout()
        
        self.label = QLabel("Entrez le texte :")
        layout.addWidget(self.label)
        
        self.text_edit = QTextEdit()
        layout.addWidget(self.text_edit)
        
        self.button = QPushButton("Calculer l'Indice de Coïncidence")
        self.button.clicked.connect(self.calculate_ic)
        layout.addWidget(self.button)
        
        self.result_label = QLabel("Indice de Coïncidence : N/A")
        layout.addWidget(self.result_label)
        
        self.setLayout(layout)
        self.setWindowTitle("Analyse IC")
    
    def calculate_ic(self):
        texte = self.text_edit.toPlainText()
        ic = calculer_indice_coincidence(texte)
        self.result_label.setText(f"Indice de Coïncidence : {ic:.6f}")

if __name__ == '__main__':
    app = QApplication(sys.argv)
    ex = ICApp()
    ex.show()
    sys.exit(app.exec_())

File: TP1_Crypto/Analyse coincidance/indice_coincidence.c
#include <stdio.h>
#include <ctype.h>
#include <string.h>

#define ALPHABET_SIZE 26

double calculer_indice_coincidence(const char *texte) {
    int frequences[ALPHABET_SIZE] = {0};
    int total_lettres = 0;
    
    // Compter les fréquences des lettres
    for (int i = 0; texte[i] != '\0'; i++) {
        char c = tolower(texte[i]);
        if (c >= 'a' && c <= 'z') {
            frequences[c - 'a']++;
            total_lettres++;
        }
    }
    
    if (total_lettres <= 1) return 0.0;
    
    // Calcul de l'indice de coïncidence
    double ic = 0.0;
    for (int i = 0; i < ALPHABET_SIZE; i++) {
        ic += (double)(frequences[i] * (frequences[i] - 1));
    }
    ic /= (double)(total_lettres * (total_lettres - 1));
    
    return ic;
}

int main() {
    char texte[1000];
    
    printf("Entrez le texte à analyser:\n");
    fgets(texte, sizeof(texte), stdin);
    
    double ic = calculer_indice_coincidence(texte);
    
    printf("Indice de coïncidence : %.6f\n", ic);
    
    return 0;
}

